clash-prelude/src/Clash/Signal/Internal.hs

```diff
@@ -1617,11 +1619,19 @@ hzToFs freq = Femtoseconds (floor ((1.0 / freq) / 1.0e-15))
 -- i.e. to calculate the clock frequency of a clock with a period of 5000 ps:
 --
 -- >>> periodToHz 5000
+-- 2.0e8
+--
+-- To get the old behaviour of this function, use type applications:
+--
+-- >>> periodToHz @Natural @(Ratio Natural) 5000
 -- 200000000 % 1
 --
 -- __NB__: This function is /not/ synthesizable
-periodToHz :: Natural -> Ratio Natural
-periodToHz period = 1.0 / (1.0e-12 * fromIntegral period)
+periodToHz :: (Integral a, Fractional b) => a -> b
+periodToHz = fromRational . toRational . periodToHzx
+ where
+  periodToHzx :: Integral a => a -> Ratio Natural
+  periodToHzx period = 1.0 / (1.0e-12 * fromIntegral period)
```

Comment:
(This is a repost from a contribution that somehow disappeared. _Very_ frustrating, I'm glad Christiaan still had an e-mail notification with it.)

Ugh, there's no call stack on the exception emitted when you pass in a negative argument:
```
*** Exception: arithmetic underflow
```
Let's handle that.
```suggestion
periodToHz :: HasCallStack => (Integral a, Fractional b) => a -> b
periodToHz period
  | period > 0
  = fromRational $ 1.0 / (fromIntegral period * 1e-12)
  | otherwise
  = withFrozenCallStack $ error "Negative or zero period"
```

You'll need to do this at the top of the file:

```diff
-import GHC.Stack                  (HasCallStack)
+import GHC.Stack                  (HasCallStack, withFrozenCallStack)
```
and your changelog entry needs to mention the changed exception for 0:
```
CHANGED: `periodToHz` now throws an `ErrorCall` with call stack when called with the argument 0 (zero), instead of a `ratioZeroDenomException :: ArithException`.
```